docs/0003-how-does-version-resolution-work-in-maven-and-gradle.md

```diff
@@ -0,0 +1,136 @@
+only loads one version of each fully-qualified class name at runtime. So, if you
+put multiple versions on the classpath, you do not get both versions at
+runtime. Java build systems handle this issue upfront by picking a single
+version for each library before the classpath is even constructed.
+
+Now let's look at the version resolution algorithms used by Gradle and
+Maven. Let's consider Gradle first because it is the simpler one to
+describe. Simply put, Gradle chooses the highest version encountered in your
+dependency graph. For the example above, Gradle picks A:10.1.
+
+<img src="assets/images/ddc-resolution-04.png" alt="Gradle example 1">
+
+Gradle applies the higher-version rule even if it overrides one of your own
+direct dependency declarations:
+
+<img src="assets/images/ddc-resolution-05.png" alt="Gradle example 2">
+
+Next consider Maven. Maven chooses the version that is closest to the root of
+the dependency graph, and if there is a tie, it chooses the version it
+encounters first. Another way to describe the algorithm is that Maven chooses
+the first version encountered when traversing the dependency graph in
+breadth-first order. Looking at the same unresolved dependency graph we used
+in Gradle example 1:
+
+<img src="assets/images/ddc-resolution-06.png" alt="Maven example 1">
+
+As you can see by comparing the resolved graphs, Maven makes the opposite
+choice. Now consider the second example, where the root has both a direct and
+indirect dependency on the same library:
+
+<img src="assets/images/ddc-resolution-07.png" alt="Maven example 2">
+
+In this case, Maven chooses A:10.0 because it is only 1 hop away from the root
+instead of 2 hops (A:10.1). In this case, Maven also makes the opposite choice
+of Gradle.
+
+It should be noted that Maven and Gradle don't always make the opposite choice.
+In the following example, the highest version is also the closest to the root,
+so Maven and Gradle both choose A:10.1.
+
+<img src="assets/images/ddc-resolution-07b.png" alt="Maven and Gradle example">
+
+## Interaction between Maven and Gradle
+
+When a library is built, its build system performs version resolution for its
+entire dependency graph. Since the build systems of the libraries in that
+dependency graph also perform their own version resolution, the versions
+selected can be different, especially when a consumer uses a different build
+system than its dependencies. Sometimes, the version selected by the consumer
+can be incompatible with the version selected by the dependency when the
+dependency is built by itself.
+
+Consider an example using Gradle example 2 as its basis, where library D depends
+on A:10.0 and C:30.0 (and thus indirectly A:10.1). Assume C depends on a feature
+added into 10.1. This means that if 10.0 is selected, C fails at runtime. From
+library D's perspective, this is fine, since Gradle chooses version 10.1.
+
+Suppose a new library E uses Maven as its build system and has a dependency on
```

Comment:
I would add the assumption that library D is now published to a Maven repository (potentially local) through Gradle's maven-publish plugin. That makes the two worlds work together.